### PR TITLE
Delayed events (Futures) / MSC4140 for call widget

### PR DESCRIPTION
**Pull request by AndrewFerr**
Depends on:
- matrix-org/matrix-widget-api#90
- matrix-org/matrix-js-sdk#4311

Signed-off-by: Andrew Ferrazzutti <andrewf@element.io>

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Comment by AndrewFerr** [comment]
I'll take this out of draft once the [dependent js-sdk PR](https://github.com/matrix-org/matrix-js-sdk/pull/4311) is merged.

**Comment by AndrewFerr** [comment]
Note that the [Static Analysis failure](https://github.com/matrix-org/matrix-react-sdk/actions/runs/10202121458/job/28225534021?pr=12714) is also [happening on develop](https://github.com/matrix-org/matrix-react-sdk/actions/runs/10201470290/job/28223379542).

File: src/stores/widgets/StopGapWidgetDriver.ts [hunk 1]
```diff
@@ -328,6 +333,93 @@ export class StopGapWidgetDriver extends WidgetDriver {
...
+            delayOpts = {
+                delay,
+                ...(parentDelayId !== null && { parent_delay_id: parentDelayId }),
+            };
+        } else if (parentDelayId !== null) {
+            delayOpts = {
+                parent_delay_id: parentDelayId,
+            };
+        } else {
+            throw new Error("Must provide at least one of delay or parentDelayId");
```
Comment (toger5) [inline src/stores/widgets/StopGapWidgetDriver.ts:382]: i think we should use the `parent_delay_id` wording here since this is an error returned to the widget and the widget will not use the camel case format.

**Review by toger5 — APPROVED** [review]
Oh forgot how easy this is since all the work is done in the widget-api repo.
This is just the matrix part!

I am positive on this, bu Just to get you confirmation: also the capability checks are done in the widget-api right?

**Comment by AndrewFerr** [comment]
> capability checks are done in the widget-api right?

Yes they are, in [src/ClientWidgetApi.ts](https://github.com/matrix-org/matrix-widget-api/blob/master/src/ClientWidgetApi.ts)

File: src/stores/widgets/StopGapWidgetDriver.ts (hunk 1 above)
Comment (AndrewFerr) [reply src/stores/widgets/StopGapWidgetDriver.ts:382]: This is a tricky one...using snake case implies that what's missing here is a field of the `send_event` widget, but what's actually missing is a parameter of an internal call within the embedded client.

If this error is hit, it means the client side of the widget-api didn't check a received `send_event` payload properly, and asked the driver to send a delayed event with no delay-related parameters, when it should have instead asked to send a non-delayed event.

In practice, this will never happen, since the widget-api indeed does all that checking, but there's no way for driver implementations to know that.  The best way to avoid all this would be to use better types on the parameter list of `sendDelayedEvent`, so that it's obvious from the driver's perspective that either one or both of the delay-related parameters will be present.

...but that's a lot of faff for something that might not even become stable, so for now I'll just leave this as-is. :slightly_smiling_face:

Comment (toger5) [reply src/stores/widgets/StopGapWidgetDriver.ts:382]: yes leave it as is. I actually was not fully aware, that this just for the driver part.

**Comment by AndrewFerr** [comment]
The latest batch of commits is just a rebase on the `develop` branch, for resolving the merge conflict on yarn.lock (where the only conflict was which version of the js-sdk to use).  All other content is the same as before.

File: src/stores/widgets/StopGapWidgetDriver.ts [hunk 2]
```diff
@@ -271,20 +276,20 @@ export class StopGapWidgetDriver extends WidgetDriver {
...
-        targetRoomId?: string,
+        targetRoomId: string | null,
     ): Promise<ISendEventDetails>;
     public async sendEvent(
         eventType: string,
         content: IContent,
-        stateKey?: string | null,
-        targetRoomId?: string,
+        stateKey: string | null = null,
+        targetRoomId: string | null = null,
```
Comment (dbkr) [inline src/stores/widgets/StopGapWidgetDriver.ts:292]: why the changes from undefined to null? I think in general intend want to move in the other direction.

File: src/stores/widgets/StopGapWidgetDriver.ts [hunk 3]
```diff
@@ -328,6 +333,93 @@ export class StopGapWidgetDriver extends WidgetDriver {
         return { roomId, eventId: r.event_id };
     }
 
+    /**
+     * @experimental Part of MSC4140 & MSC4157
+     */
```
Comment (dbkr) [inline src/stores/widgets/StopGapWidgetDriver.ts:338]: Could this have some doc on the params please? For instance, I'm at a loss as to what `parentDelayId ` might be.

Also, what would it mean to pass `null` to `delay` here?

Comment (dbkr) [reply src/stores/widgets/StopGapWidgetDriver.ts:338]: Also, this is quite a lot of params for a function. May want to consider making it an object.

File: test/stores/widgets/StopGapWidgetDriver-test.ts [hunk 4]
```diff
@@ -388,6 +391,94 @@ describe("StopGapWidgetDriver", () => {
...
+
+        it("sends delayed message events", async () => {
+            client._unstable_sendDelayedEvent.mockResolvedValue({
+                delay_id: "id",
+            });
+
+            await expect(driver.sendDelayedEvent(2000, null, EventType.RoomMessage, {})).resolves.toEqual({
+                roomId,
+                delayId: "id",
+            });
```
Comment (dbkr) [inline test/stores/widgets/StopGapWidgetDriver-test.ts:417]: Could do with an assertion that the client function is actually called in the way you think it should be called, otherwise this isn't testing what it claims to (same for the rest of these).

File: test/stores/widgets/StopGapWidgetDriver-test.ts [hunk 5]
```diff
@@ -388,6 +391,94 @@ describe("StopGapWidgetDriver", () => {
         });
     });
 
+    describe("sendDelayedEvent", () => {
+        let driver: WidgetDriver;
+        const roomId = "!this-room-id";
+
+        beforeEach(() => {
+            driver = mkDefaultDriver();
+            jest.spyOn(SdkContextClass.instance.roomViewStore, "getRoomId").mockReturnValue(roomId);
```
Comment (dbkr) [inline test/stores/widgets/StopGapWidgetDriver-test.ts:400]: Please remember to always called `restoreAllMocks()` in an `afterEach()` if you mock or spy on anything, otherwise your mocks will bleed into other tests.

**Review by dbkr — CHANGES_REQUESTED** [review]

File: src/stores/widgets/StopGapWidgetDriver.ts (hunk 2 above)
Comment (AndrewFerr) [reply src/stores/widgets/StopGapWidgetDriver.ts:292]: It's to match the signature of `WidgetApi#sendEvent`, which uses `string | null = null` for these arguments.

It also struck me as dangerous to allow `stateKey` to be `undefined` or `null` when there was only a `!== null` check on it, and felt that restricting both parameters to just `null` made things simpler.

File: src/stores/widgets/StopGapWidgetDriver.ts (hunk 3 above)
Comment (AndrewFerr) [reply src/stores/widgets/StopGapWidgetDriver.ts:338]: This implements `WidgetDriver#sendDelayedEvent`, which is present as of v1.8.0 & is where the documentation is.  I'll add a `@see` here to link to it.

Also, making this take an object would require updating the widget-api.  It doesn't take an object for the sake of avoiding adding a new type in case the whole "parent" idea gets dropped.

File: test/stores/widgets/StopGapWidgetDriver-test.ts (hunk 4 above)
Comment (AndrewFerr) [reply test/stores/widgets/StopGapWidgetDriver-test.ts:417]: Good catch, thanks!

File: test/stores/widgets/StopGapWidgetDriver-test.ts (hunk 5 above)
Comment (AndrewFerr) [reply test/stores/widgets/StopGapWidgetDriver-test.ts:400]: It turns out I don't need this after all, so I'll just get rid of it.

File: src/stores/widgets/StopGapWidgetDriver.ts (hunk 2 above)
Comment (dbkr) [reply src/stores/widgets/StopGapWidgetDriver.ts:292]: ok, it would probably be good to change the widget api at some point but that needn't be today

**Review by dbkr — APPROVED** [review]
looks good then!

**Comment by AndrewFerr** [comment]
I've rebased + squashed the commits to hide some of the churn that built up over time, and also wrote a more descriptive commit message.
